# Title: Probe Quality Control Script
# Date: 19-09-2024

# Introduction:
# This script filters out low-quality probes based on several criteria:
# - Detection p-values
# - Cross-reactive probes
# - SNP-containing probes
# - Probes from X and Y chromosomes
# - High-intensity probes
# - Low bead counts
# These steps help ensure the quality of the data for downstream analysis.

# Load necessary packages
library(minfi)
library(tidyverse)
library(ENmix)
library(maxprobes)
library(wateRmelon)

# Define file paths (customizable)
input_dir <- "../results/02_sample_QC/"
output_dir <- "../results/03_probes_QC/"
probes_dir <- "../docs/probes/"

# Load input data
load(file.path(input_dir, "sf_rgSet.RData"))
load(file.path(input_dir, "sf_MSet.RData"))
load(file.path(input_dir, "sf_grSet.RData"))

## Step 0: Analyze control probes
# This visual approach creates multiple figures and will be saves in current directory.
# The old Illumina 450k and EPIC v1 array incorparate 15 different types of
# internal control probes (total of 848 probes for 450k, 635 probes for EPICv1)
# The control plots from ENmix function plotCtrl() are similar to the control
# plots generated by Illumina GenomeStudio software.
# See Illumina Infinium HD Methylation Assay for detailed description on how
# how to interpret these control figures.
# https://www.bioconductor.org/packages/release/bioc/vignettes/ENmix/inst/doc/ENmix.html#6_Quality_Control

plotCtrl(sf.rgSet)

## Step 1: Detection p-value filtering

# The same as for samples, we want to remove probes showing high detection p-values. 
# For example the scanner can failed to read signals from some probes due to low intensities or artifacts in the array.
# Calculate detection p-values
detP <- detectionP(sf.rgSet)
detP <- detP[match(featureNames(sf.MSet), rownames(detP)), ]

# Filter probes with detection p-values higher than the threshold (cut off could be 0.01, but that is also arbitrary):
# It could be probes that have failed in one or more samples/ or probes that have failed in at least x% of the samples
# In this case we remove probes thathave detection p-value>0.01 in more than 1% of the samples: 

threshold <- 0.99 * ncol(sf.MSet)
failed.probes1 <- names(which(rowSums(detP < 0.01) < threshold))

# Get the methylation dataset and subset the failed probes (removing good probes)
failed.probes1 <- rownames(sf.MSet[failed.probes1])

print(paste("Number of poor-quality probes that have detP > 0.01 in more than 1% of the samples:",
            length(failed.probes1)))

## Step 2: Remove cross-reactive probes

# 6% of the array probes can potentially generate spurious signals because of co-hybridization 
# to alternate genomic sequences highly homologous to the intended targets." as written in the paper below:
# https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1066-1

# For EPIC V2 array: Peters et al. 2024 https://doi.org/10.1186/s12864-024-10027-5
# Have made their own manifest file where thet identified the off-target CpG sites
# list of cross-reactive probes was taken from (Additional file 8):
# https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-024-10027-5#Sec31

cross_reactive <- read.csv(file = file.path(out.dir, "cross_reactive_probes_file.csv"), header = TRUE)
rm(cross_reactive)

failed.probes2 <- unique(cross_reactive$ProbeID)
failed.probes2 <- rownames(sf.MSet[failed.probes2])
print(paste("Number of cross-reactive probes: ",
            length(failed.probes2)))

# **Note** for cross-reactive probes in EPIC V1 check: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1066-1

## Step 3: Remove probes containing SNPs

# SNPs info and MAF info is coming from dbSnp database (https://bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html#snps)
# Delete the probes that contain either a SNP at the CpG interrogation or at the single nucleotide extension site
# drop SNPs with MAF > 10 % - this is arbitrary decision

all.probes <- rownames(sf.MSet)
snp_maf_threshold <- 0.1
fun3.rgSet <- dropLociWithSnps(sf.grSet, snps = c("SBE", "CpG"), maf = snp_maf_threshold)
nosnp.probes <- rownames(fun3.rgSet)

failed.probes3 <- all.probes[!(all.probes %in% nosnp.probes)]

failed.probes3 <- rownames(sf.MSet[failed.probes3])

print(paste("Failed probes due to SNPs: ",
            length(failed.probes3)))

## Step 4: Exclude probes from X and Y chromosomes

# delete probes on X and Y chromosomes
annEPICv2 <- minfi::getAnnotation(IlluminaHumanMethylationEPICv2anno.20a1.hg38) # upload annotation
annEPICv2[1:6,1:5]

# Check whether the X and Y CpG probe is in the rownames of methylation dataset
failed.probes4 <- row.names(sf.MSet[row.names(sf.MSet) %in% annEPICv2$Name[annEPICv2$chr %in% c("chrX", "chrY")], ])
failed.probes4 <- rownames(sf.MSet[failed.probes4])

print(paste("X and Y chromosome probes: ",
            length(failed.probes4)))
# Removing the annotation data to create more working space
rm(annEPICv2)

## Step 5: Remove high-intensity probes

# From this paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4239800/
#  "For example, we have looked at the relation between signal intensities and DNA methylation measurements 
# [b-values, defined as the ratio of the methylated signal over the total signal (methylated + unmethylated) 
# and have observed that probes displaying a high average intensity (i.e. a high average of the methylated 
# and unmethylated signals) are more prone than probes displaying a lower average intensity to provide DNA 
# methylation measurements inconsistent with measurements obtained with other approaches, such as BPS (bisulfite pyrosequencing).
# They have a tendency to provide values close to 0.5, independently of their true methylation state. 
# Of note, type II Infinium probes seem to be less prone to this phenomenon."  

manifest <- getManifest(sf.rgSet)
snpProbesI <- getProbeInfo(manifest, type = "I")$Name #extract type I probes

Meth <- getMeth(sf.MSet)
Unmeth <- getUnmeth(sf.MSet)

# Median calculation
Methmed <- apply(Meth, 1, median)
Unmethmed <- apply(Unmeth, 1, median)

high_intensity_cutoff =10000
MUmed <- sqrt(Methmed * Unmethmed)
hi.MU <- names(which(MUmed > 10000)) # cutoff at MU = 10,000

# select only type 1 Porbes 
failed.probes5 <- hi.MU [hi.MU %in% snpProbesI] # select only type 1 probes

print(paste("Failed probes due to high intensity: ",
            length(failed.probes5)))

## Step 6: Filter probes with low bead counts

### Get probes  more than 5% of samples have bead numer below 3
# **Note: the Watermelon beadcount() function work with rgSet extended object.**
# Load extended rgSet
load(file = file.path(input_dir, "../rgSet_ext.RData"))

# Bead count filtering
df.bead.counts <- wateRmelon::beadcount(rgSet_ext)

# Function to calculate percentage of NAs in each row (probes)
pct.of.nas <- function(row) {
  sum(is.na(row)) / length(row) * 100
}

df.bead.counts$na_pct <- apply(df.bead.counts, 1, pct.of.nas)

# Filter out probes > 5% NAs
fun6.rgSet <- sf.MSet[which(df.bead.counts$na_pct > 5),]
failed.probes6 <- rownames(fun6.rgSet)

print(paste("Failed probes due to low bead counts: ",
            length(failed.probes6)))

## Step 7: Remove all bad quality probes and save results

failed.probes <- unique(c(
    failed.probes1,
    failed.probes2,
    failed.probes3,
    failed.probes4,
    failed.probes5,
    failed.probes6
))

## Save failed probes

write(failed.probes1, file = "../out_dir/failed_probes1_low_detection_pvalues.txt")
write(failed.probes2, file = "../out_dir/failed_probes2_crossreactive.txt")
write(failed.probes3, file = "../out_dir/failed_probes3_snpsites_at_interrogation.txt")
write(failed.probes4, file = "../out_dir/failed_probes4_xy_problematic_probes.txt")
write(failed.probes5, file = "../out_dir/failed_probes5_high_intensity_signals.txt")
write(failed.probes6, file = "../out_dir/failed_probes6_low_bead_counts.txt")
write(failed.probes, file = "../out_dir/failed_probes.txt")

print(paste("Total probes to discard: ",
            length(failed.probes6)))

## Removing probes from the dataset

# Filter the MSet by removing failed probes
r.probes.raw <- which(rownames(sf.MSet) %in% failed.probes)
pf.sf.MSet <- sf.MSet[-r.probes.raw,] # MethylSet



# Save the probe filtered and sample filtered methylation data
save(pf.sf.MSet, file = "../out_dir/pf_sf_MSet.RData")
